Upgrade to TypeScript 4.5

niik opened the pull request:
<!--
What GitHub Desktop issue does this PR address? (for example, #1234)
-->

## Description
<!--
A summary of the changes made along with any other information that would be helpful to a reviewer such as potential tradeoffs or alternative approaches you considered.
-->

This upgrades TypeScript to 4.5 while steering well clear of upgrading Webpack which is another beast entirely. In order to avoid upgrading webpack I had to use ts-loader 8.x instead of 9.x.

Additionally this does not leverage the unknown type for errors in catch blocks which we should investigate turning on separately.

## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes: no-notes

niik committed:
Clean up and prepare for TS 4.5

niik committed:
Fix TS 4.5 errors

niik committed:
Bump TS to 4.5

niik committed:
Let's see if we can get by with es2020

niik committed:
Pragmatic upgrade

niik committed:
Use ts-loader instead of awesome-typescript-loader

awesome-typescript-loader requires TS ^3

niik committed:
Abstract methods can't have implementations

niik committed:
Can't pass Element as HTMLElement

niik committed:
hello darkness my old friend

typescript no longer recognizes showModal

niik committed:
It's a thing

niik committed:
We're using ts-loader now

niik committed:
Abstract methods can't have implementations

niik committed:
Use typed instance

niik committed:
Wrong return type

niik committed:
Gotta emit in ts-loader land

niik committed:
This is why we can't have nice things

Webpack complains that it doesn't understand the emitted JS

niik committed:
These exist in TS own lib now

niik committed:
Remove more types already defined in TS lib

niik committed:
Use Electron's types for net request

headers in options work but is undocumented

Co-Authored-By: Sergio Padrino <1083228+sergiou87@users.noreply.github.com>

niik committed:
Clean up already defined types

niik committed:
:art: cleanup

niik committed:
Removed to much

tidy-dev commented on app/src/lib/globals.d.ts:
```diff
@@ -41,57 +41,6 @@ declare const __UPDATES_URL__: string
  */
 declare const __PROCESS_KIND__: 'main' | 'ui' | 'crash' | 'highlighter'
 
-/**
```
With upgrade in typescript, do these things just exist and is why we don't need them in globals.d.ts anymore?

sergiou87 replied:
Yup!

It seems like TypeScript goes a bit behind the latest API standard, and updates their [lib.dom.d.ts](https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts) later than we get new features from Electron/Chromium updates.

tidy-dev reviewed (dismissed):
✨  This seems to all make sense. Tho, I with my limited experience in this sort of thing, I am not sure I know what to look out for. :P

niik committed:
Merge branch 'development' into ts-4.5-minimal

niik committed:
Merge branch 'development' into ts-4.5-minimal

niik committed:
Upgrade typescript-eslint to a version compatible with TS 4.5

niik committed:
Bump jsdoc eslint plugin

See https://github.com/gajus/eslint-plugin-jsdoc/issues/629

niik committed:
Fix no-use-before-define lint error in tutorial-panel.tsx

niik committed:
Fix no-use-before-define lint error in tab-bar.tsx

niik committed:
Use overload-capable no-redeclare rule

niik committed:
Fix @typescript-eslint/no-unused-vars lint errors

niik committed:
Re-export it for backwards compatibility

niik committed:
Upgrade eslint-plugin-react...

in an attempts to get the no-unused-props replacement working

niik committed:
Remove eslint-plugin-babel

The no-invalid-this errors where it really shouldn't error

niik committed:
Remove react-this-binding-issue rule

We need to get rid of tslint-microsoft-contrib and this rule isn't even recommended any more https://github.com/microsoft/tslint-microsoft-contrib/issues/676#issuecomment-484881451

niik committed:
Skip the entire plugin

niik committed:
Remove tslint

niik committed:
Skip children prop

sergiou87 reviewed (dismissed):
Awesome job! Looks great to me 😄

niik committed:
Merge pull request #13807 from desktop/ts-4.5-linting

TypeScript 4.5 linting

niik commented:
Need a lil' re-review. Nothing has changed that hasn't already been reviewed though. We just merged https://github.com/desktop/desktop/pull/13807 into this branch
